cpp/src/arrow/compute/exec.cc

```diff
@@ -792,7 +802,7 @@ class ScalarExecutor : public KernelExecutorImpl<ScalarKernel> {
     preallocate_contiguous_ =
         (exec_context()->preallocate_contiguous() && kernel_->can_write_into_slices &&
          validity_preallocated_ && !is_nested(output_descr_.type->id()) &&
-         !is_dictionary(output_descr_.type->id()) &&
+         !is_dictionary(out_type_id) &&
```

Comment:
nit: if we're going to update this here, also update the usage of `output_descr_.type->id()` on the previous line?